File: backend/src/models/questionnaire.ts
import {
  arrayProp,
  instanceMethod,
  InstanceType,
  ModelType,
  prop,
  Ref,
  staticMethod,
  Typegoose,
} from "typegoose";
import { Section } from "./interfaces";

class Questionnaire extends Typegoose {
  @staticMethod
  public static findByName(
    this: ModelType<Questionnaire> & typeof Questionnaire,
    name: string,
  ) {
    return this.find({ name });
  }
  @staticMethod
  public static findByOrganization(
    this: ModelType<Questionnaire> & typeof Questionnaire,
    organizationId: string,
  ) {
    return this.find({ organizationId });
  }
  @staticMethod
  public static findByCountry(
    this: ModelType<Questionnaire> & typeof Questionnaire,
    countryId: string,
  ) {
    return this.find({ countryId });
  }
  @staticMethod
  public static findBySettlement(
    this: ModelType<Questionnaire> & typeof Questionnaire,
    settlementId: string,
  ) {
    return this.find({ settlementId });
  }
  @staticMethod
  public static findByDate(
    this: ModelType<Questionnaire> & typeof Questionnaire,
    date: string,
  ) {
    return this.find({ created: { $gte: date } });
  }
  //
  @prop({ required: true, default: [] })
  public sections!: Section[];
  //
  @prop({ required: true, index: true, trim: true, unique: true })
  public title?: string;
  //
  @prop({ required: true })
  public description?: string;

  @prop({ required: true })
  public countryName?: string;

  @prop({ required: true, index: true, trim: true })
  public countryId?: string;
  //
  @prop({ index: true, trim: true })
  public questionnaireId?: string;
  //
  @prop({ required: true, index: true, trim: true })
  public organizationId?: string;
  //
  @prop({ required: true, index: true, trim: true })
  public organizationName?: string;
  //
  @prop({ required: true, default: new Date().toISOString() })
  public created?: string;
  //
}

export default Questionnaire;

File: monitorlibrary/lib/data/questionnaire.dart
import 'package:meta/meta.dart';
import 'package:monitorlibrary/data/question.dart';
import 'package:monitorlibrary/data/section.dart';

import 'content.dart';

class Questionnaire {
  String? questionnaireId, title, name, organizationId, description, organizationName, countryId, created, countryName;
  List<Section>? sections;


  Questionnaire({required this.title, required this.name,
    required this.organizationId, required this.description, required this.organizationName,
    required this.countryId, required this.created, required this.countryName, required this.sections});

  Questionnaire.fromJson(Map data) {
    this.title = data['title'];
    this.organizationId = data['organizationId'];
    this.sections = [];
    if ( data['sections'] != null) {
      List list =  data['sections'];
      list.forEach((m) {
        this.sections!.add(Section.fromJson(m));
      });
    }
    this.description = data['description'];
    this.organizationName = data['organizationName'];
    this.countryId = data['countryId'];
    this.created = data['created'];
    this.countryName = data['countryName'];

  }
  Map<String, dynamic> toJson() {
    List mSecs = [];
    if (sections != null) {
      sections!.forEach((s) {
        mSecs.add(s.toJson());
      });
    }

    Map<String, dynamic> map = {
      'title': title,
      'organizationId': organizationId,
      'description': description,
      'sections': mSecs,
      'organizationName': organizationName,
      'created': created,
      'countryName': countryName,
      'countryId': countryId,


    };
    return map;
  }
}
